src/hotspot/share/classfile/systemDictionaryShared.cpp

```diff
@@ -329,32 +462,97 @@ bool SystemDictionaryShared::check_for_exclusion_impl(InstanceKlass* k) {
     return true;
   }
 
+  return false;
+}
+
+// Returns true if the DumpTimeClassInfo::is_excluded() is true for at least one of k's exclusion dependencies.
+bool SystemDictionaryShared::check_dependencies_exclusion(InstanceKlass* k, DumpTimeClassInfo* info) {
```

Comment:
So this checks one level of InstanceKlass* k's dependencies, and the table above collects one level of k or k's subclass's dependencies and calls this for those.

class K : extends B implements I, J verifies C {} - the above hashtable has B, I and J and C.

This function takes B - so checks B class.

class B : extends A implements intf verifies C again, why not {}

So it'd check A and intf and C again.  But when do we check A's super?

It seems like the hashtable should be created for all levels of the hierarchy, so only one bit of code should walk the hierarchy, not two.

Reply:
The ExclusionCheckCandidates table contains all (recursive) supertypes -- E.g., its `add_candidate(k)` method calls `add_candidate(k->java_super())`.

Reply:
`ExclusionCheckCandidates::add_candidate()` is a recursive function so it eventually adds all java_supers up to Object.